### PR TITLE
Bring up changes from nuttx/pci branch (2nd attempt)

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary

This is work from nuttx/pci and https://github.com/apache/nuttx/pull/1004 branches from @sonicyang and @btashton  with my cosmetic changes and squashing related commits.

Previous PR without changes from @btashton : https://github.com/apache/nuttx/pull/11570

In this PR, qemu pci code is not based on Jailhouse so the code is included.

## Testing
pcitest on qemu

### CODE REVIEW

File: include/nuttx/virt/qemu_pci.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_VIRT_QEMU_PCI_TEST_H
```
Comment (xiaoxiang781216) [inline include/nuttx/virt/qemu_pci.h:21]: __INCLUDE_NUTTX_VIRT_QEMU_PCI_H

File: drivers/virt/Make.defs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+endif
+
+# Include virt device driver build support
+#
+ifeq ($(CONFIG_VIRT),y)
+
+DEPPATH += --dep-path virt
+VPATH += :virt
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(TOPDIR)$(DELIM)drivers$(DELIM)virt}
```
Comment (xiaoxiang781216) [inline drivers/virt/Make.defs:41]:
```
${INCDIR_PREFIX}$(TOPDIR)$(DELIM)drivers$(DELIM)virt
```

File: drivers/pcie/Make.defs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+ifeq ($(CONFIG_PCIE),y)
+
+CSRCS += pcie_root.c
+
+# Include PCIE device driver build support
+
+DEPPATH += --dep-path pcie
+VPATH += :pcie
+CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" $(TOPDIR)$(DELIM)drivers$(DELIM)pcie}
```
Comment (xiaoxiang781216) [inline drivers/pcie/Make.defs]:
```
${INCDIR_PREFIX}$(TOPDIR)$(DELIM)drivers$(DELIM)pcie
```

File: drivers/pcie/pcie_root.c [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+struct pcie_dev_type_s *pci_device_types[] =
```
Comment (xiaoxiang781216) [inline drivers/pcie/pcie_root.c]:
```
static const struct pcie_dev_type_s *g_pci_device_types[] =
```

File: drivers/virt/qemu_edu.c [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+ *****************************************************************************/
+
+/*****************************************************************************
+ * Name: qemu_edu_probe
+ *
+ * Description:
+ *   Initialize device
+ *****************************************************************************/
+
+int qemu_edu_probe(FAR struct pci_bus_s *bus,
```
Comment (xiaoxiang781216) [inline drivers/virt/qemu_edu.c]: static

**Review by btashton — APPROVED** [review]

File: drivers/pci/Make.defs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+############################################################################
+# drivers/pci/Make.defs
```
Comment (xiaoxiang781216) [inline drivers/pci/Make.defs:2]: need remark

Comment (raiden00pl) [reply drivers/pci/Make.defs:2]: fixed

File: drivers/virt/qemu_edu.c [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+                              FAR struct qemu_edu_priv_s *drv_priv)
+{
+  uintptr_t  base_addr  = drv_priv->base_addr;
+  FAR void  *test_block;
+  size_t     block_size = 2048;
+  int        i;
+  uint32_t   psrand;
+  uint32_t   tx_checksum;
+  uint32_t   rx_checksum;
+  uint32_t   dev_addr   = 0x40000;
```
Comment (xiaoxiang781216) [inline drivers/virt/qemu_edu.c:278]: is it good to use a hard code address

File: include/nuttx/pci/pci.h [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+  /* Read 8, 16, 32, 64 bits data to PCI-E configuration space of device
+   * specified by dev.
+   */
+
+  CODE uint32_t (*pci_cfg_read)(FAR struct pci_dev_s *dev, int reg,
+                                int width);
+
+  /* Map address in a 32 bits bar in the memory address space */
+
+  CODE int (*pci_map_bar)(uint64_t addr, uint64_t len);
```
Comment (xiaoxiang781216) [inline include/nuttx/pci/pci.h:235]: should we support map return a different address?

File: drivers/pci/pci.c [hunk 9]
```diff
@@ -0,0 +1,832 @@
...
+  tmp_dev.bdf = PCI_BDF(bus_idx, slot_idx, dev_func);
+  vid = root_bus->ops->pci_cfg_read(&tmp_dev, PCI_CONFIG_VENDOR, 2);
+  if (vid == 0xffff)
+    {
+      return;
+    }
+
+  /* Check if this is a PCI-PCI bridge device */
+
+  sec_bus = pci_check_pci_bridge(root_bus, bus_idx, slot_idx, dev_func);
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply drivers/pci/pci.c:246]: this handles pci-pci bridges that doesn't have 0x80 bit set in the header type register (MF - multi function)

File: arch/x86_64/src/intel64/intel64_mbfb.c [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+void fb_term_initialize(void);
+#endif  /* CONFIG_MULTBOOT2_FB_TERM */
+
+void fb_clear(void);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct multiboot_fb_s fb =
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/x86_64/src/intel64/intel64_mbfb.c]: fixed

File: include/nuttx/pci/pci.h (hunk 8 above)
Comment (xiaoxiang781216) [reply include/nuttx/pci/pci.h:235]: OK

File: drivers/pci/pci.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply drivers/pci/pci.c:246]: but, the same bridge will scan twice if it has multi bit set. The right approach is moving this code after line 286(else clause).

[the remark this answers is not in the document]
Comment (raiden00pl) [reply drivers/pci/pci.c]: you are right:
```
pci_check_pci_bridge: [00:06.0] Found Bridge
pci_check_pci_bridge: [00:06.0] Found Bridge
pci_check_pci_bridge: [00:07.0] Found Bridge
pci_check_pci_bridge: [00:07.0] Found Bridge
```

fixed now:
```
pci_check_pci_bridge: [00:06.0] Found Bridge
pci_check_pci_bridge: [00:07.0] Found Bridge
```

**Comment by xiaoxiang781216** [comment]
what' about the remaining comment?

**Comment by raiden00pl** [comment]
> what' about the remaining comment?

Hidden behind Github "Load more..." button and I missed these....

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**